crates/chain/src/tx_graph.rs

```diff
@@ -304,17 +439,141 @@ impl TxGraph {
-    pub fn insert_txout_preview(&self, outpoint: OutPoint, txout: TxOut) -> Additions {
+    pub fn insert_txout_preview(&self, outpoint: OutPoint, txout: TxOut) -> Additions<A> {
         let mut update = Self::default();
         update.txs.insert(
             outpoint.txid,
-            TxNode::Partial([(outpoint.vout, txout)].into()),
+            (
+                TxNode::Partial([(outpoint.vout, txout)].into()),
+                BTreeSet::new(),
+                0,
+            ),
         );
         self.determine_additions(&update)
     }
+
+    /// Returns the resultant [`Additions`] if the `txid` is set in `anchor`.
+    pub fn insert_anchor_preview(&self, txid: Txid, anchor: A) -> Additions<A> {
+        let mut update = Self::default();
+        update.anchors.insert((anchor, txid));
+        self.determine_additions(&update)
+    }
+
+    /// Returns the resultant [`Additions`] if the `txid` is set to `seen_at`.
+    ///
+    /// Note that [`TxGraph`] only keeps track of the lastest `seen_at`.
+    pub fn insert_seen_at_preview(&self, txid: Txid, seen_at: u64) -> Additions<A> {
+        let mut update = Self::default();
+        let (_, _, update_last_seen) = update.txs.entry(txid).or_default();
+        *update_last_seen = seen_at;
+        self.determine_additions(&update)
+    }
+
+    /// Get all heights that are relevant to the graph.
+    pub fn relevant_heights(&self) -> BTreeSet<u32> {
+        self.anchors
+            .iter()
+            .map(|(a, _)| a.anchor_block().height)
+            .collect()
+    }
+
+    /// Determines whether a transaction of `txid` is in the best chain.
+    ///
+    /// TODO: Also return conflicting tx list, ordered by last_seen.
+    pub fn try_get_chain_position<C>(
+        &self,
+        chain: C,
+        txid: Txid,
+    ) -> Result<Option<ObservedIn<&A>>, C::Error>
+    where
+        C: ChainOracle,
+    {
+        let (tx_node, anchors, &last_seen) = match self.txs.get(&txid) {
+            Some((tx, anchors, last_seen)) if !(anchors.is_empty() && *last_seen == 0) => {
+                (tx, anchors, last_seen)
+            }
+            _ => return Ok(None),
+        };
+
+        for anchor in anchors {
+            if chain.is_block_in_best_chain(anchor.anchor_block())? {
+                return Ok(Some(ObservedIn::Block(anchor)));
+            }
+        }
+
+        // The tx is not anchored to a block which is in the best chain, let's check whether we can
+        // ignore it by checking conflicts!
+        let tx = match tx_node {
+            TxNode::Whole(tx) => tx,
+            TxNode::Partial(_) => {
+                // [TODO] Unfortunately, we can't iterate over conflicts of partial txs right now!
+                // [TODO] So we just assume the partial tx does not exist in the best chain :/
+                return Ok(None);
+            }
+        };
+
+        // [TODO] Is this logic correct? I do not think so, but it should be good enough for now!
```

Comment:
I think ti's ok? Remove this comment or replace it with an explanation of why we chose this algorithm.

Reply:
Fixed in 7810059ed0f23cae7dee61fe587a1c8f3f49480a